### DISCUSSION TITLE
Packaging on Windows

### DISCUSSION THREAD

**Discussion by jonathanfox5** [MahmoudAshraf97/ctc-forced-aligner#41] [Q&A] [answered]
I'm looking to replace `whisperX` in my (free to use, agpl licensed) desktop application due to `whisperX` being unmaintained. Mahmoud helpfully suggested that I check `ctc-forced-aligner` out to replace the forced alignment that `whisperX` does.

The main roadblock for implementing it is how to include the library on Windows. I'm less concerned about the users on macOS / Linux as the problem can be fixed with `homebrew` on macOS and Linux users can generally be trusted to know how to use their package manager!

On Windows, `ctc-forced-aligner` seems to require `Microsoft C++ Build Tools` to install the python package. I've thought about how to provide this to my users:
1. I could include it as part of my existing Innosetup installer which currently installs dependencies then installs the python packages using a `uv` based solution.
   - This wouldn't really work with the standard `Microsoft C++ Build Tools` web based installer, I don't want to scare less tech-savvy users away.
   - Installing silently isn't particularly an option either. Reports on the internet point to it being buggy and sometimes taking hours.
   - Offline installation is a) designed for enterprise use and b) is 3.5 GB of bulk I'd rather not add into the installer
2. For various reasons related to other packages, I _**really**_ don't want to go down the cx_freeze, etc. route.
3. I could fork and provide pre-built binaries. Not particularly elegant but it would work!

I'm leaning towards 3) but I was wondering if there was a more elegant solution?

Thanks in advance!

**Comment by MahmoudAshraf97** [comment]
Hi, prebuilt binaries are the cleanest and simplest solution, I'm willing to accept a PR for that if you're willing to contribute

**Comment by MahmoudAshraf97** [comment]
Btw, windows also faces some problems with perl because it's not bundledinstallation, so keep that in mind

**Comment by jonathanfox5** [comment]
Thanks, I'll take a look. Once I've got it working in a sane way I'll submit it as a pull request (I'm not a software engineer, it takes me a while to work out the "proper" way to do things!)

**Reply by hlevring** [reply]
@jonathanfox5 What would you plan be here, would it be to use embedded phyton and create a standalone CLI without dependencies?

**Reply by jonathanfox5** [reply]
To be honest, I haven't made any progress on this. For my project, I've made the decision to stick with whisperX for now so it will be a while (if at all) before I look into this.

**Comment by hlevring** [comment]
@MahmoudAshraf97 it looks to me like the perl dependency comes from the romanizer

But there seems to be a phyton version of that library already

https://github.com/isi-nlp/uroman

**New Python version: 1.3.1.1 (released on June 27, 2024)**
Last Perl version: 1.2.8 (released on April 23, 2021)

**Reply by MahmoudAshraf97** [reply]
I haven't tested the python implementation thoroughly tbh, so I'm reluctant to add it, one more concern is speed, whether the python implementation will be as fast as perl or not

**Reply by dperenic** [reply]
@MahmoudAshraf97 I am encountering the same issue. The package compiled successfully on Windows using VS2022, but I am now facing the Perl requirement, which is not inherently native to Windows. Integrating the Python-based alternative to uroman would be a significant improvement, as it would streamline dependencies and enhance compatibility. Given that a Python implementation already exists, it may be worthwhile to explore its feasibility in terms of performance.

**Comment by MahmoudAshraf97** [comment, marked as answer]
you can download prebuilt wheels from Actions -> latest successful workflow -> Artifacts, and then choose the one compatible with your system
